chainer/sequential.py

```diff
@@ -0,0 +1,400 @@
+                name = layer.__class__.__name__
+                param_info = '\t'
+                for param in sorted(layer.params(), key=lambda p: p.name):
+                    param_info += param.name
+                    if param._data[0] is not None:
+                        param_info += str(param._data[0].shape)
+                    else:
+                        param_info += '(None)'
+                    param_info += '\t'
+                name = name + param_info
+            elif isinstance(layer, function.Function):
+                name = layer.__class__.__name__
+            elif isinstance(layer, functools.partial):
+                name = repr(layer)
+            elif layer.__name__ == '<lambda>':
+                name = inspect.getsource(layer).strip()
+            else:
+                name = layer.__name__
+            ret += '{}\t{}\n'.format(i, name)
+        return ret
+
+    def append(self, layer):
+        self.insert(len(self), layer)
+
+    def extend(self, sequential):
+        for layer in sequential:
+            self.append(layer)
+
+    def insert(self, i, layer):
+        if not callable(layer):
+            raise ValueError(
+                'All elements of the argment should be callable. But '
+                'given {} is not callable.'.format(layer))
+
+        self._layers.insert(i, layer)
+        if isinstance(layer, link.Link):
+            self._children.insert(i, layer)
+            for i, layer in enumerate(self.children()):
+                layer.name = str(i)
+
+    def remove(self, layer):
+        if layer in self:
+            del self[self.index(layer)]
+        else:
+            raise ValueError(
+                'There is no layer object that is same as {}'.format(layer))
+
+    def remove_by_layer_type(self, type_name):
+        """Remove layers by layer type.
+
+        This method removes layers from the Sequential object by the
+        layer's class name or function name. If you want to remove a
+        :class:`~Link`, the argment ``type_name`` should be its class name,
+        e.g., :class:`~links.Linear` or :class:`~links.Convolution2D`, etc.
+        If you want to remove a :class:`~Function` class or any other callable
+        objects, ``type_name`` should be the function name, e.g., ``relu`` or
+        ``reshape``, etc.
+
+        Args:
+            type_name (str): The name of a layer you want to remove.
+
+        """
+
+        names = []
+        for layer in self:
+            if isinstance(layer, link.Link):
+                name = layer.__class__.__name__
+            else:
+                name = layer.__name__
+            names.append((name, layer))
+        for _name, _layer in names:
+            if type_name == _name:
+                self.remove(_layer)
+
+    def pop(self, i=-1):
+        layer = self._layers[i]
+        del self[i]
+        return layer
+
+    def clear(self):
+        for i, _ in enumerate(self._children):
```

Comment:
It may be good to add a TODO comment to reduce the computational cost (or fix it now if it's simple).

Reply:
Added a comment.